taichi/analysis/constexpr_propagation.cpp

```diff
@@ -0,0 +1,96 @@
+#include "taichi/ir/visitors.h"
+#include "taichi/ir/statements.h"
+
+#include <unordered_set>
+#include <functional>
+
+TLANG_NAMESPACE_BEGIN
+
+namespace {
+
+// A statement is considered constexpr in this pass, iff both its value and the
+// control flow reaching it are constant w.r.t. const_seed(...)
+
+class ConstExprPropagation : public IRVisitor {
+ public:
+  using is_const_seed_func = std::function<bool(Stmt *)>;
+
+  ConstExprPropagation(const is_const_seed_func &is_const_seed)
+      : is_const_seed_(is_const_seed) {
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  bool generic_test(Stmt *stmt) {
+    if (is_const_seed_(stmt)) {
+      const_stmts_.insert(stmt);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  void visit(Stmt *stmt) override {
+    generic_test(stmt);
+  }
+
+  bool is_inferred_const(Stmt *stmt) {
+    // Note: every statement that tests true by "is_const_seed_" should have
+    // already been included in const_stmts_.
+    return const_stmts_.find(stmt) != const_stmts_.end();
+  };
+
+  void visit(UnaryOpStmt *stmt) override {
+    if (generic_test(stmt))
+      return;
+    if (is_inferred_const(stmt->operand)) {
+      const_stmts_.insert(stmt);
+    }
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    if (generic_test(stmt))
+      return;
+    if (is_inferred_const(stmt->lhs) && is_inferred_const(stmt->rhs)) {
+      const_stmts_.insert(stmt);
+    }
+  }
+
+  void visit(IfStmt *stmt) override {
+    if (is_inferred_const(stmt->cond)) {
```

Comment:
I would suggest adding a comment here to remind that the control flow reaching it is constant w.r.t. const_seed(...).

Maybe TODO: What about `TernaryOpStmt`s and `RangeForStmt`s with `is_inferred_const(stmt->begin) && is_inferred_const(stmt->end)`?

Reply:
Good point. I'll add `TernaryOpStmt` in this PR and leave `RangeForStmt` as TODO. How to define a variable is `constexpr` in the latter case with looping, is not yet clear to me.